File: java/src/test/java/site/markhenrick/recreational/adventofcode/y2022/Day02Test.java
package site.markhenrick.recreational.adventofcode.y2022;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.MethodSource;
import site.markhenrick.recreational.common.TestUtil;

import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.params.provider.Arguments.arguments;
import static site.markhenrick.recreational.adventofcode.y2022.Day02.Shape.*;

public class Day02Test {
	private static final int WIN = 2;
	private static final int LOSS = 0;
	private static final int DRAW = 1;

	private static final String SAMPLE_INPUT = """
A Y
B X
C Z
""";

	@ParameterizedTest
	@CsvSource({
		"ROCK, 1",
		"PAPER, 2",
		"SCISSORS, 3",
	})
	void shapeScore(Day02.Shape shape, int expectedScore) {
		assertThat(shape.score()).isEqualTo(expectedScore);
	}

	static Stream<Arguments> fight() {
		return Stream.of(
			arguments(ROCK, ROCK, DRAW),
			arguments(ROCK, PAPER, LOSS),
			arguments(ROCK, SCISSORS, WIN),
			arguments(PAPER, ROCK, WIN),
			arguments(PAPER, PAPER, DRAW),
			arguments(PAPER, SCISSORS, LOSS),
			arguments(SCISSORS, ROCK, LOSS),
			arguments(SCISSORS, PAPER, WIN),
			arguments(SCISSORS, SCISSORS, DRAW)
		);
	}

	@ParameterizedTest
	@MethodSource
	void fight(Day02.Shape ours, Day02.Shape theirs, int expectedScore) {
		assertThat(ours.fight(theirs)).isEqualTo(3 * expectedScore);
	}

	@ParameterizedTest
	@MethodSource("fight")
	void findMatch(Day02.Shape expected, Day02.Shape theirs, int desiredOutcome) {
		assertThat(Day02.Shape.findMatch(theirs, desiredOutcome)).isEqualTo(expected);
	}

	static Stream<Arguments> part1Test() {
		return Stream.of(
			Arguments.of(SAMPLE_INPUT, 15),
			arguments(TestUtil.getResourceAsString("AoC/input/2022/day02.txt"), 11063)
		);
	}

	@ParameterizedTest
	@MethodSource
	void part1Test(String input, int expected) {
		assertThat(Day02.part1(input)).isEqualTo(expected);
	}

	static Stream<Arguments> part2Test() {
		return Stream.of(
			Arguments.of(SAMPLE_INPUT, 12),
			arguments(TestUtil.getResourceAsString("AoC/input/2022/day02.txt"), 10349)
		);
	}

	@ParameterizedTest
	@MethodSource
	void part2Test(String input, int expected) {
		assertThat(Day02.part2(input)).isEqualTo(expected);
	}
}

File: kotlin/src/test/kotlin/site/markhenrick/recreational/adventofcode/y2025/Day02Test.kt
package site.markhenrick.recreational.adventofcode.y2025

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.junit.jupiter.params.provider.ValueSource
import site.markhenrick.recreational.getChallengeInput
import java.util.stream.Stream

class Day02Test {
    @ParameterizedTest
    @MethodSource
    fun part1(input: String, expected: Long) {
        assertThat(Day02.solvePart1(input)).isEqualTo(expected)
    }

    @ParameterizedTest
    @MethodSource
    fun part2(input: String, expected: Long) {
        assertThat(Day02.solvePart2(input)).isEqualTo(expected)
    }

    @ParameterizedTest
    @ValueSource(longs = [
        1,
        12,
        123,
        1221,
        101,
    ])
    fun validIds1(id: Long) {
        assertThat(Day02.isValid1(id)).isTrue
    }

    @ParameterizedTest
    @ValueSource(longs = [
        11,
        22,
        99,
        1010,
        1188511885,
        222222,
        446446,
        38593859,
    ])
    fun invalidIds1(id: Long) {
        assertThat(Day02.isValid1(id)).isFalse
    }

    @ParameterizedTest
    @ValueSource(longs = [
        1,
        12,
        123,
        1221,
        101,
    ])
    fun validIds2(id: Long) {
        assertThat(Day02.isValid2(id)).isTrue
    }

    @ParameterizedTest
    @ValueSource(longs = [
        11,
        22,
        99,
        111,
        999,
        1010,
        1188511885,
        222222,
        446446,
        38593859,
        565656,
        824824824,
        2121212121
    ])
    fun invalidIds2(id: Long) {
        assertThat(Day02.isValid2(id)).isFalse
    }

    companion object {
        private val sampleInput = "11-22,95-115,998-1012,1188511880-1188511890,222220-222224,1698522-1698528,446443-446449,38593856-38593862,565653-565659,824824821-824824827,2121212118-2121212124"
        private val challengeInput = getChallengeInput(2025, 2)

        @JvmStatic
        fun part1(): Stream<Arguments> = Stream.of(
            Arguments.of(sampleInput, 1227775554),
            Arguments.of(challengeInput, 34826702005)
        )

        @JvmStatic
        fun part2(): Stream<Arguments> = Stream.of(
            Arguments.of(sampleInput, 4174379265),
            Arguments.of(challengeInput, 43287141963L)
        )
    }
}
